opentelemetry-sdk/src/logs/growable_array.rs

```diff
@@ -0,0 +1,369 @@
+{
+    fn new(source: GrowableArray<T, INITIAL_CAPACITY>) -> Self {
+        Self {
+            iter: Self::get_iterator(source),
+        }
+    }
+
+    fn get_iterator(
+        source: GrowableArray<T, INITIAL_CAPACITY>,
+    ) -> std::iter::Chain<
+        std::iter::Take<std::array::IntoIter<T, INITIAL_CAPACITY>>,
+        std::vec::IntoIter<T>,
+    > {
+        if source.additional.is_none() || source.additional.as_ref().unwrap().is_empty() {
+            source
+                .initial
+                .into_iter()
+                .take(source.count)
+                .chain(Vec::<T>::new())
+        } else {
+            source
+                .initial
+                .into_iter()
+                .take(source.count)
+                .chain(source.additional.unwrap())
+        }
+    }
+}
+
+impl<T: Default + Clone + PartialEq, const INITIAL_CAPACITY: usize> Iterator
+    for GrowableArrayIntoIter<T, INITIAL_CAPACITY>
+{
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iter.next()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::logs::growable_array::{
+        GrowableArray, DEFAULT_INITIAL_VEC_CAPACITY, DEFAULT_MAX_STACK_CAPACITY,
+    };
+    use opentelemetry::logs::AnyValue;
+    use opentelemetry::Key;
+
+    /// A struct to hold a key-value pair and implement `Default`.
+    #[derive(Clone, Debug, PartialEq)]
+    struct KeyValuePair(Key, AnyValue);
+
+    impl Default for KeyValuePair {
+        fn default() -> Self {
+            KeyValuePair(Key::from_static_str(""), AnyValue::String("".into()))
+        }
+    }
+
+    #[test]
+    fn test_push_and_get() {
+        let mut collection = GrowableArray::<i32>::new();
+        for i in 0..15 {
+            collection.push(i);
+        }
+        for i in 0..15 {
+            assert_eq!(collection.get(i), Some(&(i as i32)));
+        }
+    }
+
+    #[test]
+    fn test_len() {
+        let mut collection = GrowableArray::<i32>::new();
+        for i in 0..15 {
+            collection.push(i);
+        }
+        assert_eq!(collection.len(), 15);
+    }
+
+    #[test]
+    fn test_into_iter() {
+        let mut collection = GrowableArray::<i32>::new();
+        for i in 0..15 {
+            collection.push(i);
+        }
+        let mut iter = collection.into_iter();
+        for i in 0..15 {
+            assert_eq!(iter.next(), Some(i));
+        }
+        assert_eq!(iter.next(), None);
+    }
+
+    #[test]
+    fn test_ref_iter() {
+        let mut collection = GrowableArray::<i32>::new();
+        for i in 0..15 {
+            collection.push(i);
+        }
+        let iter = collection.iter();
+        let mut count = 0;
+        for value in iter {
+            assert_eq!(*value, count);
+            count += 1;
+        }
+        assert_eq!(count, 15);
+    }
+
+    #[test]
+    fn test_key_value_pair_storage_growable_array() {
+        let mut collection = GrowableArray::<KeyValuePair>::new();
```

Comment:
It might be better to test `GrowableArray` with the type that actually gets used by `LogRecord` which is `Option<(Key, AnyValue)>` instead of `struct KeyValuePair(Key, AnyValue)`.

Reply:
yes, have updated the tests for this. Thanks once again :)